### PR TITLE
refactor(Core/Cache): Move the GlobalPlayerCache to its own class

### PR DESCRIPTION
**Pull request by Nyeriah**
## Changes Proposed:
-  Move the global cache to its own class
-  Fixes change faction not working due to mail cache overflow
- Implement commands to check and debug the global cache (cache info, cache delete and cache refresh)
- Cache refresh allows reseting the cache data for the given player if something wrong happens
- Delete caches will only refresh correctly the next time players relog (use it for debugging only)

## Issues Addressed:

- Closes 

## Tests Performed:

- Tested in-game
- 

## How to Test the Changes:

1. Send mails/delete mails
2. use the .cache info $playerName command
3. Check if the count displayed matches & you can faction change normally

## Known Issues and TODO List:

- [ ] merge https://github.com/azerothcore/mod-azerothshard/pull/15/files
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Nyeriah** [comment]
Lots of things could use testing
- Creating arena teams/inviting to arena teams
- Auction house bidding/buyouts 
- Sending mails
- Commands...

**Comment by Winfidonarleyan** [comment]
Very good. I wanted to do the same thing

File: src/server/game/Cache/CharacterCache.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    ObjectGuid Guid;
+    std::string Name;
+    uint32 AccountId;
+    uint8 Class;
+    uint8 Race;
+    uint8 Sex;
+    uint8 Level;
+    uint8 MailCount;
+    ObjectGuid::LowType GuildId;
+    uint32 ArenaTeamId[3];
```
Comment (Winfidonarleyan) [inline src/server/game/Cache/CharacterCache.h:37]: `std::array` maybe?

File: src/server/game/Cache/CharacterCache.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    ObjectGuid GroupGuid;
+};
+
+class AC_GAME_API CharacterCache
+{
+    public:
+        CharacterCache();
+        ~CharacterCache();
```
Comment (Winfidonarleyan) [inline src/server/game/Cache/CharacterCache.h:45]:
```suggestion
        CharacterCache() = default;
        ~CharacterCache() = default;
```

**Comment by Winfidonarleyan** [comment]
- [x] Work

**Comment by Nyeriah** [comment]
https://github.com/azerothcore/mod-azerothshard/pull/15/files

File: src/server/game/Cache/CharacterCache.cpp [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    std::unordered_map<std::string, CharacterCacheEntry*> _characterCacheByNameStore;
+}
+
+CharacterCache::CharacterCache()
+{
+}
+
+CharacterCache::~CharacterCache()
+{
+}
```
Comment (Winfidonarleyan) [inline src/server/game/Cache/CharacterCache.cpp:40]: delete this and in `.h` file can
```cpp
CharacterCache() noexcept = default;
~CharacterCache() noexcept = default;
```

Comment (Nyeriah) [reply src/server/game/Cache/CharacterCache.cpp]: Thanks, I had gotten that error and had no time to look into it, it should be fine now

**Comment by Nyeriah** [comment]
Okay, I took care of the last suggestion. Let me know if anything else is needed.

**Comment by IntelligentQuantum** [comment]
https://github.com/azerothcore/mod-azerothshard/blob/master/modules/mod-playerstats/src/scripts/PlayerStats.cpp#L28

**Comment by Nyeriah** [comment]
> https://github.com/azerothcore/mod-azerothshard/pull/15/files

**Review by Winfidonarleyan — APPROVED** [review]

**Comment by Winfidonarleyan** [comment]
Can merge

**Comment by patou01** [comment]
@nyeriah , I believe this has broken the core? Anyone working on a fix?

**Comment by Nyeriah** [comment]
> @Nyeriah , I believe this has broken the core? Anyone working on a fix?

https://github.com/azerothcore/azerothcore-wotlk/pull/9244

**Pull request merged**